### PR TITLE
#15 - Productionize response from ChatGPT

### LINKED ISSUE #13: JSON format the response from OpenAI
For the bee inspections, we want to be able to analyze the input and output a JSON object to then visualize the insights of the bee hive health over time. 

When we make the call to openAI we want the response to be formatted.

### LINKED ISSUE #15: productionize JSON response from ChatGPT
get ChatGPT to consistently output a JSON of the values in the beekeeping report for rendering

transcript -> system prompt -> JSON

also investigate processing the transcript to reduce unnecessary pauses or hmms to save space

### PR DESCRIPTION
**Pull request by idjevm**
### What does this pull request aim to achieve?
- Clean up / refactor and restructure the swarm-rest-api as well as continue the work from @Clue355  on issue 13 
- Refactored to use Java Records to be more easily deserialize the response so later it will be easy to persist when we start doing that work
- created a `responseJsonStructure` which should be what we expect from ChatGPT.
- Refactored from Carlos' work to use HttpClient as it is simpler

Now we are getting the follow response -> 
image

I know a few things can be changed, but those can be on follow up small issues, and small PRs, so we can move past this step of getting the first setup done. 

To look at my changes directly look into [my commit here](https://github.com/LittleTechBookClub/swarm-rest-api/commit/2d17a8b8336e2b9e3044c13821ee31a920c0cbfb)

### What is left to do from this? 
- Fix up unit test
- handle all TODOs left on the comments
- Some more refactoring
- more refactoring and other changes that can be done in following issues / PRs.

_Please be concise and link any related issue(s):_
closes #15
closes dup #14 
closes #13 from @Clue355 

_Please do **NOT** submit this PR if you have not done the following:_

- [x] The name of this PR is `#issueNumber - short description`
- [x] I have filled the issue number in the above lines to link any related issue(s)
- [x] I have checked that no one else is working on similar changes.
- [x] The name of this branch is: **`issueNumber-short-description`**, and that the branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging**
- [x] There are no merge conflicts.
- [x] On the right-hand side of this PR, I have linked the appropriate projects under `Projects`, requested at least 2 reviewers under `Reviewers`, and I have assigned this PR to myself & others needed under `Assignees`

### CODE REVIEW

**Comment by Clue355** [comment]
looks like the test failed in the project. Edit that to make it work

**Review by Clue355 — CHANGES_REQUESTED** [review]
test is failing

**Comment by idjevm** [comment]
> test is failing

I actually left a comment that I will not be handling tests on this PR but will need to make a separate PR for following TODOs  & tests.

File: src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java
Comment (MelisaIm) [inline src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java:30]: @lindentree I think we randomly considered using an API key per user (to offload the cost to the end user). Did you get the chance to look into that? (cc @Clue355 @idjevm)

File: src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java
Comment (MelisaIm) [inline src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java:46]: Yes, let's do that ^

File: src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java
Comment (MelisaIm) [inline src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java:64]: 👏🏽

File: src/main/java/com/ltbc/swarmrestapi/service/beesinfo/Food.java
Comment (MelisaIm) [inline src/main/java/com/ltbc/swarmrestapi/service/beesinfo/Food.java:7]: For bee food data, maybe we can also track blooming patterns for flowers in the area 🤔 using a different API

File: src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java
Comment (lindentree) [reply src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java:30]: I think we’d have to set up a client side thing for the user to enter their chatgpt key and authorize its usage for our third party app.

File: src/test/java/com/littletech/swarmrestapi/OpenAIApiControllertest.java
Comment (lindentree) [inline src/test/java/com/littletech/swarmrestapi/OpenAIApiControllertest.java:26]: OpenAIApiControllertest -> OpenAIApiControllerTest

**Review by lindentree — CHANGES_REQUESTED** [review]

File: src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java
Comment (MelisaIm) [reply src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java:30]: @lindentree Let's bring it up with Moni and Shamique :) I'll post on Slack

File: src/test/java/com/littletech/swarmrestapi/OpenAIApiControllertest.java
Comment (idjevm) [reply src/test/java/com/littletech/swarmrestapi/OpenAIApiControllertest.java:26]: Ah that's probably why it is broken lol I didn't look at the tests at all but this makes sense

**Review by Clue355 — APPROVED** [review]

File: src/main/java/com/ltbc/swarmrestapi/service/beesinfo/Food.java
Comment (idjevm) [reply src/main/java/com/ltbc/swarmrestapi/service/beesinfo/Food.java:7]: We could but we can put a ticket in the backlog to enhance these APIs later. I think for the first example / first API this one should be fine and from here we can expand and create more APIs

Comment (idjevm) [reply src/main/java/com/ltbc/swarmrestapi/service/beesinfo/Food.java:7]: Created #18  from your comment, we can refine it later.

File: src/test/java/com/littletech/swarmrestapi/OpenAIApiControllertest.java
Comment (idjevm) [reply src/test/java/com/littletech/swarmrestapi/OpenAIApiControllertest.java]: I ended up fixing unit tests here so this should be resolved now - changed a few things though but this PR should be shippeable now

File: src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java
Comment (idjevm) [reply src/main/java/com/ltbc/swarmrestapi/service/SwarmRestApiService.java:46]: This I will handle in another PR / refactor later on

**Review by lindentree — APPROVED** [review]
LGTM! Maybe it would be better to have @fsantamaria1’s PR merged first though, to spread out the contributions.

**Comment by idjevm** [comment]
> LGTM! Maybe it would be better to have @fsantamaria1’s PR merged first though, to spread out the contributions.

Yup, I asked him to do that so I am going to hold off for his PR to be merged before merging this one.

**Pull request merged**
